src/metabase/api/database.clj

```diff
@@ -1058,35 +1058,42 @@
 
 (api/defendpoint GET "/:id/schemas"
   "Returns a list of all the schemas found for the database `id`"
-  [id include_editable_data_model include_hidden]
+  [id include_editable_data_model include_hidden include_empty]
   {id                          ms/PositiveInt
    include_editable_data_model [:maybe ms/BooleanString]
-   include_hidden              [:maybe ms/BooleanString]}
+   include_hidden              [:maybe ms/BooleanString]
+   include_empty               [:maybe ms/BooleanString]}
   (let [include_editable_data_model (Boolean/parseBoolean include_editable_data_model)
         include_hidden              (Boolean/parseBoolean include_hidden)
-        filter-schemas (fn [schemas]
-                         (if include_editable_data_model
-                           (if-let [f (u/ignore-exceptions
-                                       (classloader/require 'metabase-enterprise.advanced-permissions.common)
-                                       (resolve 'metabase-enterprise.advanced-permissions.common/filter-schema-by-data-model-perms))]
-                             (map :schema (f (map (fn [s] {:db_id id :schema s}) schemas)))
-                             schemas)
-                           (filter (partial can-read-schema? id) schemas)))]
-    (if include_editable_data_model
-      (api/check-404 (t2/select-one Database id))
-      (api/read-check Database id))
-    (->> (t2/select-fn-set :schema Table
-                           :db_id id :active true
-                           (merge
-                            {:order-by [[:%lower.schema :asc]]}
-                            (when-not include_hidden
-                              ;; a non-nil value means Table is hidden -- see [[metabase.models.table/visibility-types]]
-                              {:where [:= :visibility_type nil]})))
-         filter-schemas
-         ;; for `nil` schemas return the empty string
-         (map #(if (nil? %) "" %))
-         distinct
-         sort)))
+        include_empty               (Boolean/parseBoolean include_empty)]
+    (if include_empty
+      ;; include_empty is mutually exclusive with the include_editable_data_model and include_hidden params
+      (do (api/check-superuser)
+          (let [db (api/check-404 (t2/select-one Database id))]
+            (driver/all-schemas (:engine db) db)))
```

Comment:
You're right. I created a new endpoint, `GET api/database/:id/syncable_schemas`, and renamed `driver/all-schemas` to `driver/syncable-schemas`. This better reflects the exact meaning of the set of schemas returned by the endpoint, because it should actually be the list of schemas that if tables are in, they will be synced.